pkg/frontend/middleware/audit.go

```diff
@@ -0,0 +1,134 @@
+package middleware
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/log"
+	"github.com/Azure/ARO-RP/pkg/util/log/audit"
+)
+
+// Audit generates an audit log based on the request, caller, resource and
+// correlation data found in a HTTP request. It depends on the 'Log' middleware
+// to populate the request context with data that it needs.
+func Audit(env env.Core, entry *logrus.Entry) func(http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// skip requests to the 'operationsstatus' endpoints
+			if strings.Contains(r.URL.Path, "operationsstatus") {
+				h.ServeHTTP(w, r)
+				return
+			}
+
+			callerIdentity, callerType, correlationID, requestID := callerRequestData(r)
+			targetResourceName, targetResourceType := targetResourceData(entry, r)
+
+			auditEntry := audit.NewEntry(entry.Logger)
+			auditEntry = auditEntry.WithFields(logrus.Fields{
+				audit.EnvKeyAppID:             audit.SourceRP,
+				audit.EnvKeyCloudRole:         audit.CloudRoleRP,
+				audit.EnvKeyCorrelationID:     correlationID,
+				audit.EnvKeyEnvironment:       env.Environment().Name,
+				audit.EnvKeyHostname:          env.Hostname(),
+				audit.EnvKeyLocation:          env.Location(),
+				audit.PayloadKeyCategory:      audit.CategoryResourceManagement,
+				audit.PayloadKeyOperationName: fmt.Sprintf("%s %s", r.Method, r.URL.Path),
+				audit.PayloadKeyRequestID:     requestID,
+				audit.PayloadKeyCallerIdentities: []audit.CallerIdentity{
+					{
+						CallerIdentityType:  callerType,
+						CallerIdentityValue: callerIdentity,
+						CallerIPAddress:     r.RemoteAddr,
+					},
+				},
+				audit.PayloadKeyTargetResources: []audit.TargetResource{
+					{
+						TargetResourceName: targetResourceName,
+						TargetResourceType: targetResourceType,
+					},
+				},
+			})
+
+			defer func() {
+				resultType, resultDescription := resultData(w)
+				auditEntry.WithFields(logrus.Fields{
+					audit.PayloadKeyResult: audit.Result{
+						ResultType:        resultType,
+						ResultDescription: resultDescription,
+					},
+				}).Info("audit event")
+			}()
+
+			h.ServeHTTP(w, r)
+		})
+	}
+}
+
+func callerRequestData(r *http.Request) (
+	callerIdentity string,
+	callerType string,
+	correlationID string,
+	requestID string) {
+
+	callerIdentity = r.UserAgent()
+	callerType = audit.CallerIdentityTypeApplicationID
+
+	if v := r.Context().Value(ContextKeyCorrelationData); v != nil {
+		if correlationData, ok := v.(*api.CorrelationData); ok {
```

Comment:
The only case in which this code should fail to run is if the Audit middleware is not run after the Log middleware.
If that happens, this code will fail silently and audit logs will be affected.
Write `correlationData := r.Context().Value(middleware.ContextKeyCorrelationData).(*api.CorrelationData)` here, so that this code will panic if there is a refactor which breaks it.
We have already taken a similar decision elsewhere in the codebase, e.g. in _putOrPatchOpenShiftCluster().

Reply:
> so that this code will panic

although i don't like the idea of letting my code panic because someone else decided to make some incompatible changes elsewhere, it makes sense in this case because we are making a deliberate design decision to couple the audit middleware to the log middleware, on the basis that they should log the same data. i will include this change.

fwiw, i believe the [`recover.Panic()` in `frontend.Run()`](https://github.com/Azure/ARO-RP/blob/078d3b25fb826575a3a7f1c6e5c61e69b0216654/pkg/frontend/frontend.go#L283) recovers the panic, before exiting, so it's not as scary as it seems in prod.